api/cryptocom/client.go

```diff
@@ -0,0 +1,276 @@
+func (c *Connection) Type() string {
+	if c.IsPrivate {
+		return "private"
+	}
+
+	return "public"
+}
+
+type HTTPClient interface {
+	Post(url, contentType string, body io.Reader) (resp *http.Response, err error)
+	Get(url string) (resp *http.Response, err error)
+}
+
+type Connection struct {
+	Endpoint  string
+	IsPrivate bool
+	Transport
+	sync.Mutex
+}
+
+type Client struct {
+	publicConn  Connection
+	privateConn Connection
+	wsRootURL   string
+	restRootURL string
+	key         string
+	secret      string
+	privateSubs []string
+	publicSubs  []string
+	httpClient  HTTPClient
+	outbox      chan Response
+	LogFunc     LogFunc
+	wg          sync.WaitGroup
+}
+
+// New returns a pointer of Client struct
+func New(wsRootURL, restRootURL, key, secret string) *Client {
+	return &Client{
+		key:         key,
+		secret:      secret,
+		wsRootURL:   wsRootURL,
+		restRootURL: restRootURL,
+		outbox:      make(chan Response),
+		privateSubs: make([]string, 0),
+		publicSubs:  make([]string, 0),
+		httpClient:  &http.Client{},
+		LogFunc:     defaultLogFunc,
+	}
+}
+
+// Connect instansiate WS Connections
+func (c *Client) Connect() error {
+	publicWsEndpoint := c.wsRootURL + marketEndpoint
+	privateWsEndpoint := c.wsRootURL + userEndpoint
+
+	err := c.createConnection(publicWsEndpoint, false)
+	if err != nil {
+		return err
+	}
+
+	err = c.createConnection(privateWsEndpoint, true)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(3 * time.Second) // Cryptocom requires this sleep.
+	c.authenticate()
+
+	return nil
+}
+
+func (c *Client) Listen() <-chan Response {
+	go c.readConnection(c.publicConn)
+	go c.readConnection(c.privateConn)
+	return c.outbox
+}
+
+func (c *Client) Shutdown() {
+	c.privateConn.Close()
+	c.publicConn.Close()
+	c.wg.Wait()
+	close(c.outbox)
+}
+
+func (c *Client) createConnection(endpoint string, isPrivate bool) error {
+	conn, _, err := websocket.DefaultDialer.Dial(endpoint, http.Header{})
+	if err != nil {
+		return err
+	}
+
+	cnx := Connection{Endpoint: endpoint, IsPrivate: isPrivate, Transport: conn}
+
+	if isPrivate {
+		c.privateConn = cnx
+	} else {
+		c.publicConn = cnx
+	}
+
+	return nil
+}
+
+func (c *Client) readConnection(cnx Connection) {
+	defer c.wg.Done()
+	c.wg.Add(1)
+	c.LogFunc("Start listening connection ... %s", cnx.Endpoint)
+	for {
+		_, m, err := cnx.ReadMessage()
+		if err != nil {
+			c.LogFunc("error on read message in %s cnx\nError message: %s\n", cnx.Type(), err.Error())
+			if isClosedCnxError(err) {
+				c.LogFunc("Stop reading from %s cnx. Connection closed\n", cnx.Type())
+				return
+			}
+			for {
+				conn, _, err := websocket.DefaultDialer.Dial(cnx.Endpoint, http.Header{})
+				if err != nil {
+					continue
+				}
+
+				time.Sleep(3 * time.Second) // Cryptocom requires this sleep
+				newCnx := Connection{Endpoint: cnx.Endpoint, IsPrivate: cnx.IsPrivate, Transport: conn}
+
+				if newCnx.IsPrivate {
+					c.privateConn = newCnx
+					c.authenticate()
+					if len(c.privateSubs) > 0 {
```

Comment:
maybe put this if inside subscribe

Reply:
I didn't get you.
In which `subscribe` do you want to put it?

Reply:
c.subscribePrivateChannels, c.subscribePublicChannels

Is it possible to use one subscribe here?

Reply:
I don't think it's possible